include/openssl/hmac.h

```diff
@@ -151,6 +151,80 @@ OPENSSL_EXPORT int HMAC_CTX_copy_ex(HMAC_CTX *dest, const HMAC_CTX *src);
 OPENSSL_EXPORT void HMAC_CTX_reset(HMAC_CTX *ctx);
 
 
+// Precomputed key functions
+
+// HMAC_MAX_PRECOMPUTED_KEY_SIZE is the largest precomputed key size, in bytes.
+#define HMAC_MAX_PRECOMPUTED_KEY_SIZE (2 * (EVP_MAX_MD_CHAINING_LENGTH))
+
+// HMAC_set_precomputed_key_export sets the context |ctx| to allow export of the
+// precomputed key using HMAC_get_precomputed_key. On entry, HMAC_CTX must have
+// been initialized via HMAC_Init_*, and neither HMAC_Update nor HMAC_Final
+// must have been called after the last HMAC_Init_ex. It returns one on success
+// and zero on programmer error.
+// After a successful call to HMAC_set_precomputed_key_export, HMAC_Update and
+// HMAC_Final will fail.
+//
+// Note: The main reason for this function is to satisfy FIPS assertion AS09.16,
+// since HMAC_get_precomputed_key returns key material (i.e., a CSP in NIST
+// terminology).
+OPENSSL_EXPORT int HMAC_set_precomputed_key_export(HMAC_CTX *ctx);
```

Comment:
I like better the second option (`HMAC_get_precomputed_key` can just set `out_len` to the correct value when out is NULL.)
I can implement it if you think it's better than having a separate function.

Reply:
Thinking more about it, I think there is a concrete difference between having a separate function to get the size or not:

* with a separate function (like in the current code), there is no restriction when the function to get the size can be called (as long as the md has been initialized)
* when merging this function with `HMAC_set_precomputed_key_export` or `HMAC_get_precomputed_key`, the size can only be retrieved in a specific state (i.e., just after initialization or just after setting the key export flag respectively).

Having a separate function gives a bit more of flexibility. But I don't know if it really matters to users of the library.

Reply:
>  If it's given a non-NULL out, it checks that out_len is not less than a certain minimum (otherwise an error is returned)

What is the purpose of testing that out_len is "not less than a certain minimum (otherwise an error is returned)"?
I tried to look at a few functions with a `*out_len` in AWS-LC and none of them checked `out_len` before setting it to the actual output length.

Reply:
This is an example I found https://github.com/aws/aws-lc/blob/main/crypto/fipsmodule/evp/p_hkdf.c#L123.
It's a good idea also to put `OPENSSL_PUT_ERROR` as in that example where it makes sense to let the user know what went wrong.

Reply:
I've modified the behavior to match the above.